lib/std/crypto/tls/cipher.zig

```diff
@@ -0,0 +1,1029 @@
+
+            // get padding len from last padding byte
+            const padding_len = plaintext[plaintext.len - 1] + 1;
+            if (plaintext.len < mac_len + padding_len) return error.TlsDecryptError;
+            // split plaintext into cleartext and mac
+            const cleartext_len = plaintext.len - mac_len - padding_len;
+            const cleartext = plaintext[0..cleartext_len];
+            const mac = plaintext[cleartext_len..][0..mac_len];
+
+            // write ad to the buf
+            var ad = additionalData(self.decrypt_seq, rec.content_type, cleartext_len);
+            @memcpy(buf[0..ad.len], &ad);
+            const mac_msg = buf[0 .. ad.len + cleartext_len];
+            self.decrypt_seq +%= 1;
+
+            // calculate expected mac and compare with received mac
+            var expected_mac: [mac_len]u8 = undefined;
+            Hmac.create(&expected_mac, mac_msg, &self.decrypt_secret);
+            if (!std.mem.eql(u8, &expected_mac, mac))
+                return error.TlsBadRecordMac;
+
+            return .{ rec.content_type, cleartext };
+        }
+    };
+}
+
+// xor lower 8 iv bytes with seq
+fn ivWithSeq(comptime nonce_len: usize, iv: [nonce_len]u8, seq: u64) [nonce_len]u8 {
+    var res = iv;
+    const buf = res[nonce_len - 8 ..];
+    const operand = std.mem.readInt(u64, buf, .big);
+    std.mem.writeInt(u64, buf, operand ^ seq, .big);
+    return res;
+}
+
+pub const additional_data_len = record.header_len + @sizeOf(u64);
+
+fn additionalData(seq: u64, content_type: proto.ContentType, payload_len: usize) [additional_data_len]u8 {
+    const header = record.header(content_type, payload_len);
+    var seq_buf: [8]u8 = undefined;
+    std.mem.writeInt(u64, &seq_buf, seq, .big);
+    return seq_buf ++ header;
+}
+
+// Cipher suites lists. In the order of preference.
+// For the preference using grades priority and rules from Go project.
+// https://ciphersuite.info/page/faq/
+// https://github.com/golang/go/blob/73186ba00251b3ed8baaab36e4f5278c7681155b/src/crypto/tls/cipher_suites.go#L226
+pub const cipher_suites = struct {
+    pub const tls12_secure = if (crypto.core.aes.has_hardware_support) [_]CipherSuite{
+        // recommended
+        .ECDHE_ECDSA_WITH_AES_128_GCM_SHA256,
+        .ECDHE_ECDSA_WITH_AES_256_GCM_SHA384,
+        .ECDHE_ECDSA_WITH_CHACHA20_POLY1305_SHA256,
+        // secure
+        .ECDHE_RSA_WITH_AES_128_GCM_SHA256,
+        .ECDHE_RSA_WITH_AES_256_GCM_SHA384,
+        .ECDHE_RSA_WITH_CHACHA20_POLY1305_SHA256,
+    } else [_]CipherSuite{
+        // recommended
+        .ECDHE_ECDSA_WITH_CHACHA20_POLY1305_SHA256,
+        .ECDHE_ECDSA_WITH_AES_128_GCM_SHA256,
+        .ECDHE_ECDSA_WITH_AES_256_GCM_SHA384,
+
+        // secure
+        .ECDHE_RSA_WITH_CHACHA20_POLY1305_SHA256,
+        .ECDHE_RSA_WITH_AES_128_GCM_SHA256,
+        .ECDHE_RSA_WITH_AES_256_GCM_SHA384,
+    };
+    pub const tls12_week = [_]CipherSuite{
```

Comment:
I think you meant "weak" instead of "week"? Same thing with some of the comments below.